rfcs/0006-transaction.md

```diff
@@ -0,0 +1,366 @@
+```
+
+Let's say the current transaction is `txn4`, it should check conflicts with `txn2` and `txn3` during the commit, because `txn2` and `txn3` are committed after `txn4` started. During the conflict checking, `txn4`'s Read & Write keys should not conflict with the Write keys of `txn2` and `txn3`.
+
+It does not need to check conflicts with `txn1`, because `txn1` is committed before `txn4` started. This could tell us when to GC the entry in the `recent_committed_txns` deque: when ALL the running transactions' sequence number are bigger than the committed sequence number of the entry, then the entry could be GCed.
+
+Please note that `read_keys` are represented as a set of `Range<Bytes>`. This allows us to efficiently track ranges of keys that were read during the transaction, and ensure Serializable Snapshot Isolation (SSI) on range conflicts. For example, if transaction A reads a range `["key01", "key10"]` and transaction B writes to `"key05"` which falls within that range, we can detect this conflict and abort one of the transactions to maintain serializability.
+
+This requires we track the original write keys in the `TransactionState`, instead of the key fingerprints like Badger does. We could reuse the buffers of the original keys in WriteBatch to reference the write keys with slices of `Bytes`.
+
+The non-transactioned writes operations should also be tracked in the `TransactionState` and `recent_committed_txns` for conflict checking. They could be regarded as a transaction that only contains a single write operation, and the start sequence number of this transaction is same as the committed sequence number.
+
+This approach can be adjusted to support SI as well by only tracking the keys that are being written to. When user set the isolation level to SI, the `read_keys` could always be set as empty. The GC mechanism to release inactive transactions is still the same as the SSI.
+
+#### GC & Expiration
+
+As described above, the entries in `recent_committed_txns` could be released whenever ALL the running transactions' sequence number are bigger than the committed sequence number of the entry. 
+
+We have to track the running transactions's sequence number in the `Oracle`.
+
+Also, it's less make sense to allow a transaction to be running for 12 hours, the `recent_committed_txns` will also keep growing for such 12 hours, and the GC will not be able to release the entries. This is a potential problem that the memory usage of the `Oracle` will keep growing, and possible lead to OOM.
+
+To mitigate this, we should allow users to set a default expiration time for all of the transactions & snapshots, and the transactions & snapshots will be considered as expired if it's not finished within the expiration time.
+
+```rust
+struct Oracle {
+    running_txns: HashSet<u64>,
+    next_seq: u64,
+    discard_seq: u64,
+    recent_committed: Deque<TransactionState>>,
+}
+
+struct TransactionState {
+    started_time: u64,
+    expired_time: u64,
```

Comment:
would the expire time be set when the transaction is started?

Reply:
yes, it's not an `Option<>`, it's set on started